store/store_action.go

```diff
@@ -564,3 +579,52 @@ func findRev(needle snap.Revision, haystack []snap.Revision) bool {
 	}
 	return false
 }
+
+func reportFetchAssertionsError(res *snapActionResult, assertq AssertionQuery) error {
+	// prefer to report the most unexpected error:
+	// * errors not referring to an assertion (no valid type/primary-key)
+	// are more unexpected than
+	// * errors referring to a precise assertion that are not not-found
+	// themselves more unexpected than
+	// * not-found errors
+	errIdx := -1
+	errl := res.ErrorList
+	carryingRef := func(i int) bool {
+		aType := asserts.Type(errl[i].Type)
+		return aType != nil && len(errl[i].PrimaryKey) == len(aType.PrimaryKey)
+	}
+	for i, ent := range errl {
```

Comment:
Nitpick, the priority checks felt a bit hard to follow. Took the liberty of tweaking this a bit locally to make code a bit simpler:

```
diff --git a/store/store_action.go b/store/store_action.go
index 76a3cb1b86..bed8d780a1 100644
--- a/store/store_action.go
+++ b/store/store_action.go
@@ -589,28 +589,32 @@ func reportFetchAssertionsError(res *snapActionResult, assertq AssertionQuery) e
        // * not-found errors
        errIdx := -1
        errl := res.ErrorList
-       carryingRef := func(i int) bool {
-               aType := asserts.Type(errl[i].Type)
-               return aType != nil && len(errl[i].PrimaryKey) == len(aType.PrimaryKey)
+       carryingRef := func(ent *errorListEntry) bool {
+               aType := asserts.Type(ent.Type)
+               return aType != nil && len(ent.PrimaryKey) == len(aType.PrimaryKey)
+       }
+       prio := func(ent *errorListEntry) int {
+               if !carryingRef(ent) {
+                       return 2
+               }
+               if ent.Code != "not-found" {
+                       return 1
+               }
+               return 0
        }
        for i, ent := range errl {
-               withRef := carryingRef(i)
-               if withRef && ent.Code == "not-found" {
-                       if errIdx == -1 {
-                               errIdx = i
-                       }
-               } else if withRef {
-                       if errIdx == -1 || errl[errIdx].Code == "not-found" {
-                               errIdx = i
-                       }
-               } else {
-                       if errIdx == -1 || carryingRef(errIdx) {
-                               errIdx = i
-                       }
+               if errIdx == -1 {
+                       errIdx = i
+                       continue
+               }
+               prioOther := prio(&errl[errIdx])
+               prioThis := prio(&ent)
+               if prioThis > prioOther {
+                       errIdx = i
                }
        }
        rep := errl[errIdx]
-       if carryingRef(errIdx) {
+       if carryingRef(&rep) {
                ref := &asserts.Ref{Type: asserts.Type(rep.Type), PrimaryKey: rep.PrimaryKey}
                var err error
                if rep.Code == "not-found" {
```

Reply:
Thank you, that looks better indeed. Fell free to push to this, anyway I think we need another master merge because of the debian issue.

Reply:
pushed